Fix an infinite loop of GarbageCollectionThread

withsmilo opened the pull request:
* Changed list
  * TaskExecutionThread's worker calls try_pop() to spin.
  * GarbageCollectionThread's worker calls wait_pop() to block.
* related issue: #468

withsmilo committed:
Fix an infinite loop of GarbageCollectionThread

* Changed list
  * TaskExecutionThread's worker calls try_pop() to spin.
  * GarbageCollectionThread's worker calls wait_pop() to block.

AmplabJenkins commented:
Can one of the admins verify this patch?

dcrankshaw commented:
jenkins ok to test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1389/
Test FAILed.

dcrankshaw commented on src/libclipper/include/clipper/threadpool.hpp:
```diff
@@ -325,11 +335,18 @@ class FixedSizeThreadPool : public ThreadPool {
 
  private:
   size_t queue_id_ = 1;
+  bool is_block_worker_ = true;
```
Rather than hardcoding this to one, can you make it a constructor argument, and then set it to `true` in `GarbageCollectionThreadpool::getThreadPool` where we call the `FixedSizeThreadpool` constructor? Whether the worker should block or not is a result of the semantics of the GarbageCollector, not a fundamental part of the FixedSizeThreadpool semantics.

withsmilo replied:
Good review. I will try to modify it.

dcrankshaw reviewed (changes requested):
Thanks for tracking this down! Just had a minor comment for improving the usability of `ThreadPool` in the future.

Commit:
Transform is_block_worker from class variable to constructor argument

Commit:
Fix check_container_activity bug

QueryFrontend sometimes crashed and caused the critical segment fault.
I googled this [StackOverflow post](https://stackoverflow.com/questions/38468844/erasing-elements-from-unordered-map-in-a-loop) and modified RPCService::check_container_activity() function and the crash was disappeared finally.

Commit:
Remove model info when all containers of the model are removed

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1438/
Test FAILed.

Commit:
Run format_code.sh

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1439/
Test FAILed.

Commit:
Run run_unittests.sh

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1440/
Test FAILed.

dcrankshaw committed:
Merge branch 'develop' into fix_infinite_thread_loop

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1442/
Test PASSed.

dcrankshaw committed:
Merge branch 'develop' into fix_infinite_thread_loop

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1446/
Test FAILed.

simon-mo committed:
Format Code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1447/
Test FAILed.

simon-mo committed:
Fix variable name

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1448/
Test PASSed.

dcrankshaw committed:
Merge branch 'develop' into fix_infinite_thread_loop

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1449/
Test PASSed.

dcrankshaw reviewed (approved):
lgtm
